ydb/core/kqp/provider/rewrite_io_utils.cpp

```diff
@@ -0,0 +1,169 @@
+    }
+
+    TExprNode::TPtr queryGraph;
+    if (!CompileExpr(*queryAst.Root, queryGraph, ctx, nullptr, nullptr)) {
+        return nullptr;
+    }
+
+    return queryGraph;
+}
+
+bool ContainsNullNode(const TExprNode::TPtr& root) {
+    return static_cast<bool>(FindNode(root, [](const TExprNode::TPtr& node) {
+        return !node;
+    }));
+}
+
+void AddChild(const TExprNode::TPtr& parent, const TExprNode::TPtr& newChild) {
+    auto childrenToChange = parent->ChildrenList();
+    childrenToChange.emplace_back(newChild);
+    parent->ChangeChildrenInplace(std::move(childrenToChange));
+}
+
+void ChangeChild(const TExprNode::TPtr& parent, ui32 index, const TExprNode::TPtr& newChild) {
+    Y_ENSURE(parent->ChildrenSize() > index);
+
+    auto childrenToChange = parent->ChildrenList();
+    childrenToChange[index] = newChild;
+    parent->ChangeChildrenInplace(std::move(childrenToChange));
+}
+
+TExprNode::TPtr FindSavedQueryGraph(const TExprNode::TPtr& carrier) {
+    if (carrier->ChildrenSize() == 0) {
+        return nullptr;
+    }
+    auto lastChild = carrier->Children().back();
+    return lastChild->IsCallable(QueryGraphNodeSignature) ? lastChild->ChildPtr(0) : TExprNode::TPtr();
+}
+
+void SaveQueryGraph(const TExprNode::TPtr& carrier, TExprContext& ctx, const TExprNode::TPtr& payload) {
+    AddChild(carrier, ctx.NewCallable(payload->Pos(), QueryGraphNodeSignature, {payload}));
+}
+
+void InsertExecutionOrderDependencies(const TExprNode::TPtr& queryGraph, const TExprNode::TPtr& worldBefore) {
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&worldBefore](const TExprNode::TPtr& node) {
+            if (node->ChildrenSize() > 0 && node->Child(0)->IsWorld()) {
+                ChangeChild(node, 0, worldBefore);
+            }
+            return true;
+        }
+    );
+}
+
+bool CheckTopLevelness(const TExprNode::TPtr& candidateRead, const TExprNode::TPtr& queryGraph) {
+    THashSet<TExprNode::TPtr> readsInCandidateSubgraph;
+    VisitExpr(candidateRead, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        if (node->IsCallable(ReadName)) {
+            readsInCandidateSubgraph.emplace(node);
+        }
+        return true;
+    });
+
+    return !FindNode(queryGraph, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        return node->IsCallable(ReadName) && !readsInCandidateSubgraph.contains(node);
+    });
+}
+
+TExprNode::TPtr FindTopLevelRead(const TExprNode::TPtr& queryGraph) {
+    const TExprNode::TPtr* lastReadInTopologicalOrder = nullptr;
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&lastReadInTopologicalOrder](const TExprNode::TPtr& node) {
+            if (node->IsCallable(ReadName)) {
+                lastReadInTopologicalOrder = &node;
+            }
+            return true;
+        }
+    );
+
+    if (!lastReadInTopologicalOrder) {
+        return nullptr;
+    }
+
+    Y_ENSURE(CheckTopLevelness(*lastReadInTopologicalOrder, queryGraph),
```

Comment:
Not sure if this is a good idea. This message is for developers, not for the user. However, I would like to check this assumption in the code somehow.

Reply:
On the other hand, if I messed up in choosing the correct Read! node in this function, then the user will get an even more cryptic error message:

"Failed to execute callable with name: ResWrite!, you possibly used cross provider/cluster operations or pulled not materialized result in refselect mode"

This message appears whenever KQP failed to substitute some ResWrite! node to a TKiExecDataQuery! node, which usually happens when there are nodes with side-effects (like Read!) not seen on the path by the first child from the root of the query graph (see [CheckTx](https://github.com/ydb-platform/ydb/blob/c288b66eb979d88e47d5c141f1fe9f853336e08f/ydb/core/kqp/provider/yql_kikimr_opt_build.cpp#L658) function).